test/editor/roundtrip-test.ts

```diff
@@ -0,0 +1,162 @@
+            ["code blocks containing backticks", "```\nfoo ->`x`\nbar\n```"],
+            ["nested formatting", "a<del>b **c _d_ e** f</del>g"],
+            ["an ordered list", "A\n\n1. b\n2. c\n3. d\nE"],
+            ["an unordered list", "A\n\n- b\n- c\n- d\nE"],
+            ["f code block followed by text after a blank line", "```A\nfoo(bar).baz();\n\n3\n```\n\nB"],
+            ["just a code block", "```\nfoo(bar).baz();\n\n3\n```"],
+            ["inline code", "there's no place `127.0.0.1` like"],
+            ["nested quotations", "saying\n\n> > foo\n\n> NO\n\nis valid"],
+            ["quotations", "saying\n\n> NO\n\nis valid"],
+            ["links", "click [this](http://example.com/)!"],
+        ])('%s', async (_name, markdown) => {
+            expect(await roundTripMarkdown(markdown)).toEqual(markdown);
+        });
+
+        test.skip.each([
+            // Removes trailing spaces
+            ["a code block followed by newlines", "```\nfoo(bar).baz();\n\n3\n```\n\n"],
+            // Adds a space after the code block
+            ["a code block surrounded by text", "```A\nfoo(bar).baz();\n\n3\n```\nB"],
+            // Adds a space before the list
+            ["an unordered list directly preceded by text", "A\n- b\n- c\n- d\nE"],
+            // Adds a space before the list
+            ["an ordered list directly preceded by text", "A\n1. b\n2. c\n3. d\nE"],
+            // Adds and removes spaces before the nested list
+            ["nested unordered lists", "A\n- b\n- c\n    - c1\n    - c2\n- d\nE"],
+            // Adds and removes spaces before the nested list
+            ["nested ordered lists", "A\n\n1. b\n2. c\n    1. c1\n    2. c2\n3. d\nE"],
+            // Adds and removes spaces before the nested list
+            ["nested mixed lists", "A\n\n1. b\n2. c\n    - c1\n    - c2\n3. d\nE"],
+            // Backslashes get doubled
+            ["backslashes", "C:\\Program Files"],
+            // Deletes the whitespace
+            ['newlines with trailing and leading whitespace', "hello \n world"],
+            // Escapes the underscores
+            ["underscores within a word", "abso_fragging_lutely"],
+            // Includes the trailing text into the quotation
+            // https://github.com/vector-im/element-web/issues/22341
+            ["quotations without separating newlines", "saying\n> NO\nis valid"],
+            // Removes trailing and leading whitespace
+            ["quotations with trailing and leading whitespace", "saying \n\n> NO\n\n is valid"],
+        ])('%s', async (_name, markdown) => {
+            expect(await roundTripMarkdown(markdown)).toEqual(markdown);
+        });
+
+        it('styling, but * becomes _', async function() {
+            expect(await roundTripMarkdown("**bold** and *emphasised*"))
+                .toEqual("**bold** and _emphasised_");
+        });
+    });
+
+    describe('HTML messages should round-trip if they contain', function() {
```

Comment:
same comments as md roundtrip, essentially.

Reply:
Done in 40f27f232a